src/crimson/os/seastore/extent_placement_manager.h

```diff
@@ -783,18 +797,48 @@ class ExtentPlacementManager {
     struct reserve_io_result_t {
       bool reserve_inline_success = true;
       bool reserve_main_ool_success = true;
+      bool reserve_cold_ool_success = true;
 
-      reserve_io_result_t(std::tuple<bool, bool> t)
+      reserve_io_result_t(std::tuple<bool, bool, bool> t)
         : reserve_inline_success(std::get<0>(t)),
-          reserve_main_ool_success(std::get<1>(t)) {}
+          reserve_main_ool_success(std::get<1>(t)),
+          reserve_cold_ool_success(std::get<2>(t))
+      {}
 
       bool is_successful() const {
         return reserve_inline_success &&
-          reserve_main_ool_success;
+          reserve_main_ool_success &&
+          reserve_cold_ool_success;
+      };
+    };
+
+    struct cleaner_usage_t {
+      main_ool_usage_t main_ool_usage;
+      cold_ool_usage_t cold_ool_usage;
+    };
+
+    struct reserve_cleaner_result_t {
+      bool reserve_main_ool_success;
+      bool reserve_cold_ool_success;
+
+      reserve_cleaner_result_t()
+        : reserve_main_ool_success(true),
+          reserve_cold_ool_success(true)
+      {}
+
+      reserve_cleaner_result_t(std::tuple<bool, bool> t)
+        : reserve_main_ool_success(std::get<0>(t)),
+          reserve_cold_ool_success(std::get<1>(t))
+      {}
+
+      bool is_successful() const {
+        return reserve_main_ool_success &&
+          reserve_cold_ool_success;
       };
     };
```

Comment:
TBH I still perfer the plain version proposed in https://github.com/ceph/ceph/pull/47974#discussion_r1061292389 due to readability:
* The reservation must not introduce deadlock as described in https://github.com/ceph/ceph/pull/47974#discussion_r973964649: the calling relationship from `try_reserve_io()`, `try_reserve_cleaner()` to `try_reserve_cold()` is explicit that lower-level reservations are the subset of the higher-level reservations.
* The resource reservation must be symmetric to the release: this is explict in both method signatures and implementations as follows:
```
abort_cold_usage()
try_reserve_cold()
abort_cleaner_usage()
try_reserve_cleaner()
abort_io_usage()
try_reserve_io()
```
* The variadic template, inheritance and tuple don't look simpler to me, and the heap allocations (`unique_ptr`) seem unnecessary.

* `do_background_cycle()` can become more readable as a result:
```c++
seastar::future<> BackgroundProcess::do_background_cycle()
{
  assert(is_ready());
  maybe_update_eviction_mode();

  bool should_trim = trimmer->should_trim();
  bool proceed_trim = false;
  std::size_t trim_cycle_size = trimmer->get_trim_size_per_cycle();
  // ...
  cleaner_usage_t trim_usage{trim_cycle_size, has_cold_tier() ? trim_cycle_size : 0};
  if (should_trim) {
    auto result = try_reserve_cleaner(trim_usage);
    if (result.is_successful()) {
      proceed_trim = true;
    }
  }

  if (proceed_trim) {
    return trimmer->trim(
    ).finally([this, trim_usage] {
      abort_cleaner_usage(trim_usage, {true, true});
    });
  } else {
    bool should_clean_main =
      main_cleaner->should_clean_space() ||
      should_do_fast_eviction() ||
      (should_trim && !proceed_trim);
    bool proceed_clean_main = false;
    std::size_t main_cold_usage = has_cold_tier() ?
      main_cleaner->get_reclaim_size_per_cycle() : 0;
    if (should_clean_main) {
      proceed_clean_main = try_reserve_cold(main_cold_usage);
    }

    bool proceed_clean_cold = false;
    if (has_cold_tier() &&
        (cold_cleaner->should_clean_space() ||
         (should_clean_main && !proceed_clean_main))) {
      proceed_clean_cold = true;
    }

    ceph_assert(proceed_clean_main || proceed_clean_cold);
    return seastar::when_all(
      [this, proceed_clean_main, main_cold_usage] {
        if (!proceed_clean_main) {
          return seastar::now();
        }
        return main_cleaner->clean_space(
        ).handle_error(
          crimson::ct_error::assert_all{
            "do_background_cycle encountered invalid error in main clean_space"
          }
        ).finally([this, main_cold_usage] {
          abort_cold_usage(main_cold_usage);
        });
      },
      [this, proceed_clean_cold] {
        if (!proceed_clean_cold) {
          return seastar::now();
        }
        return cold_cleaner->clean_space(
        ).handle_error(
          crimson::ct_error::assert_all{
            "do_background_cycle encountered invalid error in cold clean_space"
          }
        );
      }
    ).discard_result();
  }
}
```

Reply:
OK, I get it, such changes will be reverted in next push.

Reply:
Probably good to add comments about the reservation deadlock.